istioctl/pkg/rbac/upgrade.go

```diff
@@ -0,0 +1,267 @@
+		}
+	}
+	return nil
+}
+
+// upgradeServiceRole simply removes the `services` field for the serviceRolePolicy.
+func (ug *Upgrader) upgradeServiceRole(serviceRolePolicy model.Config) model.Config {
+	serviceRoleSpec := serviceRolePolicy.Spec.(*rbacproto.ServiceRole)
+	for _, rule := range serviceRoleSpec.Rules {
+		if rule.Methods == nil && rule.Paths == nil && rule.Constraints == nil {
+			// If `services` is the only field, we need to create `methods = ["*"]`
+			rule.Methods = []string{"*"}
+		}
+		rule.Services = nil
+	}
+	return serviceRolePolicy
+}
+
+// createAuthorizationPolicyFromRoleBinding creates AuthorizationPolicy from the given ServiceRoleBinding.
+// In particular:
+// * For each ServiceRoleBinding, create an AuthorizationPolicy
+//   * If field `user` exists, change it to use `names`
+//	 * If field `group` exists, change it to use `groups`
+//   * Change `roleRef` to use `role`
+//   * Create a list of binding with one element (serviceRoleBinding) and use workload selector.
+func (ug *Upgrader) createAuthorizationPolicyFromRoleBinding(serviceRoleBinding model.Config) (model.Config, error) {
+	bindingSpec := serviceRoleBinding.Spec.(*rbacproto.ServiceRoleBinding)
+	for _, subject := range bindingSpec.Subjects {
+		if subject.User != "" {
+			subject.Names = []string{subject.User}
+			subject.User = ""
+		}
+		if subject.Group != "" {
+			subject.Groups = []string{subject.Group}
+			subject.Group = ""
+		}
+		if bindingSpec.RoleRef != nil {
+			bindingSpec.Role = bindingSpec.RoleRef.Name
+			bindingSpec.RoleRef = nil
+		}
+	}
+	authzPolicy := rbacproto.AuthorizationPolicy{}
+	authzPolicy.Allow = []*rbacproto.ServiceRoleBinding{bindingSpec}
+	workloadLabels, err := ug.getAndAddRoleNameToWorkloadLabelMapping(bindingSpec.Role, serviceRoleBinding.Namespace)
+	if err != nil {
+		return model.Config{}, err
+	}
+	if len(workloadLabels) > 0 {
+		authzPolicy.WorkloadSelector = &rbacproto.WorkloadSelector{Labels: workloadLabels}
+	}
+	authzPolicyConfig := model.Config{
+		ConfigMeta: model.ConfigMeta{
+			Type:      model.AuthorizationPolicy.Type,
+			Group:     serviceRoleBinding.Group, // model.AuthorizationPolicy.Group is "rbac", but we need "rbac.istio.io".
+			Version:   model.AuthorizationPolicy.Version,
+			Name:      serviceRoleBinding.Name,
+			Namespace: serviceRoleBinding.Namespace,
+		},
+		Spec: &authzPolicy,
+	}
+	return authzPolicyConfig, nil
+}
+
+func (ug *Upgrader) getAndAddRoleNameToWorkloadLabelMapping(roleName, namespace string) (WorkloadLabels, error) {
+	workloadLabels := WorkloadLabels{}
+	if _, found := ug.RoleNameToWorkloadLabels[roleName]; !found {
+		// If RoleNameToWorkloadLabels does not have roleName key yet, try fetching from Kubernetes apiserver.
+		serviceRole := ug.IstioConfigStore.Get(model.ServiceRole.Type, roleName, namespace)
+		if serviceRole == nil {
+			return nil, fmt.Errorf("cannot find ServiceRole named %s in namespace %s", roleName, namespace)
```

Comment:
nit `cannot find ServiceRole %s in namespace %s`

Reply:
Done.